### PR TITLE
feat(rules): add no console log rule

### PR DESCRIPTION
**Pull request by 0xIchigo**
The goal of this PR is to address the [this issue](https://github.com/denoland/deno/issues/20187) presented in the main Deno GitHub repository, which proposes the implementation of a `no-console` rule similar to that of [eslint](https://eslint.org/docs/latest/rules/no-console). In order to add such a rule to `deno lint` in the main Deno repository, we need a rule defined in the linter itself.

This PR achieves that goal by adding in a `NoConsoleLog` lint rule that checks whether a `console.log()` statement is in the specified code

### CODE REVIEW

File: docs/rules/no_console_log.md
Comment (lino-levan) [inline docs/rules/no_console_log.md:9]:
```suggestion
```typescript
```

File: docs/rules/no_console_log.md
Comment (lino-levan) [inline docs/rules/no_console_log.md:14]:
```suggestion
function log() { console.log("Log"); }
```

File: docs/rules/no_console_log.md
Comment (lino-levan) [inline docs/rules/no_console_log.md:19]:
```suggestion
```typescript
```

File: docs/rules/no_console_log.md
Comment (lino-levan) [inline docs/rules/no_console_log.md:26]:
```suggestion
function log_error(message: string) {
```

File: docs/rules/no_console_log.md
Comment (lino-levan) [inline docs/rules/no_console_log.md:27]:
```suggestion
    console.warn(message);
```

**Review by lino-levan — CHANGES_REQUESTED** [review]

**Comment by 0xIchigo** [comment]
Damn, thank you for the suggested changes @lino-levan! My brain has been in Rust-mode for the last little while and I didn't even realize I put the code examples in Rust instead of TypeScript :tired_face:

The suggested changes have been implemented with [this commit](https://github.com/denoland/deno_lint/pull/1189/commits/1cbd2df0271c499b82920fa11f4a525b670364ea)

File: docs/rules/no_console_log.md
Comment (lino-levan) [inline docs/rules/no_console_log.md:22]: not sure what these are doing here

Comment (0xIchigo) [reply docs/rules/no_console_log.md]: Sorry, I can take those out - they were just to show that any other line of code that does not contain `console.log` is valid. Fixed with [this commit](https://github.com/denoland/deno_lint/pull/1189/commits/f21a23064e0fb5195390d33fa4307cf305c6a74f)

File: docs/rules/no_console_log.md
Comment (marvinhagemeister) [inline docs/rules/no_console_log.md]: I feel like putting a negative word like "guilty" is a bit much for official docs, given that this is usually done unintentionally. Maybe we an say something like: "Oftentimes, developers accidentally commit `console.log` statements, ..."

**Review by marvinhagemeister — APPROVED** [review]
LGTM, I leave it to @dsherret to merge 👍

**Comment by 0xIchigo** [comment]
Thank you for the suggestions @marvinhagemeister! I agree with them and have fixed with [this commit](https://github.com/denoland/deno_lint/pull/1189/commits/81b4a23cd8193245dd7fd54b22da48bdef37c960). LGTM 🔥

**Comment by bartlomieju** [comment]
I'm not convinced this should be in the recommended set - this means that any CLI tool that wants to output to console will spawn a plethora of errors. Maybe we should add it by default to the `fresh` set and other users can enable this rule as they see fit?

**Comment by marvinhagemeister** [comment]
@bartlomieju good point. Let's just add this rule without adding it to any preset. I don't want this to enable by default for `fresh` users but I'd like to opt-in to that for the fresh codebase itself.

**Comment by 0xIchigo** [comment]
Gentle bump @bartlomieju - is there anything else to add in order to make this an optional rule, not added to any preset by default, or is adding the functionality to the `rules` directory enough?

**Comment by lev-blit** [comment]
maybe this is for a future PR, but I find myself committing `console.debug` a lot as well (perhaps even more so than `console.log`)...

perhaps it would be useful to be able to customize the behavior of this like it's possible in eslint's [no-console rule](https://eslint.org/docs/latest/rules/no-console#options) to either disallow all `console` functions or to specify which are allowed (`error` for example).

**Comment by 0xIchigo** [comment]
@lev-blit yeah, that wouldn't be too hard to implement! I don't mind opening up another PR to tackle that issue once this one is merged. The nice part of how the code is structured is that in the implementation for `NoConsoleLogHandler` we have the following `if` check: `if prop_ident.sym().as_ref() == "log"` - you can simply swap out the `"log"` for `"debug"` in an entirely new rule, or modify the code to respond with different messages and errors based on whether you have a `console.log()` or a `console.debug()`

**Review by dsherret — APPROVED** [review]
LGTM, though I think this should align with eslint. I'm going to merge this PR as-is, then do a follow-up PR in a few minutes to convert it to be like eslint and remove it from the recommended set.

**Pull request merged**
